src/Sarif/Writers/SarifConsolidator.cs

```diff
@@ -0,0 +1,259 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Sarif.Writers
+{
+    /// <summary>
+    ///  SarifConsolidator is used to reduce the size of Sarif objects and the serialized log.
+    ///  It removes common redundant properties, uses Run-wide collections to avoid writing duplicate
+    ///  values, and can remove portions of Results entirely.
+    /// </summary>
+    public class SarifConsolidator
+    {
+        private Run _run;
+        private Dictionary<ThreadFlowLocation, int> _uniqueThreadFlows;
+
+        public int TotalThreadFlowLocations { get; private set; }
+        public int UniqueThreadFlowLocations => _uniqueThreadFlows.Count;
+
+        public int TotalLocations { get; private set; }
+        public int UniqueLocations { get; private set; }
+
+        public int? MessageLengthLimitBytes { get; set; }
+        public bool RemoveCodeFlows { get; set; }
+        public bool RemoveRelatedLocations { get; set; }
+        public bool RemoveGraphs { get; set; }
+        public bool RemoveGraphTraversals { get; set; }
+        public bool RemoveStacks { get; set; }
+        public bool RemoveWebRequests { get; set; }
```

Comment:
>RemoveWebRequests [](start = 20, length = 17)

Similar to Larry's point above not sure how flexibility to trim web request data v. response data makes sense (certainly we could add this granularity in future if need be). This all relates to a conversation we had the other day, it would be helpful, given the combinatorics of insert/removal/transform implied by this very complex format to attempt to shape the API around higher-level use cases. 'Zap all graph related things' or 'Zap all web analysis related things' is an example of a small step towards this. We may eventually benefit from an ever higher-level concept such as 'remove all data related to the detailed investigation of an analysis result'. That's really the heart of the problem here, i.e., do we want a log file that neatly summarizes the analysis or do we need one that can drive an end-to-end evaluation/disposition,.

Reply:
I've consolidated Graphs and GraphTraversals. I expect users who want to preserve log meaning to leave all of these options false. These are for when the user must remove content to make the log small enough to interpret. I thought about combining requests and responses but thought a user might want to keep requests (likely small) to confirm what was called but not responses (likely big) if they found they couldn't afford the file size.